app/Foliage/Pages.hs

```diff
@@ -50,29 +51,32 @@ data AllPackagesPageEntry = AllPackagesPageEntry
     allPackagesPageEntryTimestamp :: UTCTime,
     allPackagesPageEntryTimestampPosix :: POSIXTime,
     allPackagesPageEntrySource :: PackageVersionSource,
-    allPackagesPageEntryRevision :: Maybe RevisionSpec
+    allPackagesPageEntryLatestRevisionTimestamp :: Maybe UTCTime
   }
   deriving stock (Generic)
   deriving (ToJSON) via MyAesonEncoding AllPackagesPageEntry
 
-makeAllPackagesPage :: UTCTime -> FilePath -> [PackageVersionMeta] -> Action ()
+makeAllPackagesPage :: UTCTime -> FilePath -> [PreparedPackageVersion] -> Action ()
 makeAllPackagesPage currentTime outputDir packageVersions = do
   let packages =
-        sortOn allPackagesPageEntryPkgId
-          $ map
-            ( ( \PackageVersionMeta {pkgId, pkgSpec = PackageVersionSpec {packageVersionTimestamp, packageVersionRevisions, packageVersionSource}} ->
-                  AllPackagesPageEntry
-                    { allPackagesPageEntryPkgId = pkgId,
-                      allPackagesPageEntryTimestamp = fromMaybe currentTime packageVersionTimestamp,
-                      allPackagesPageEntryTimestampPosix = utcTimeToPOSIXSeconds (fromMaybe currentTime packageVersionTimestamp),
-                      allPackagesPageEntrySource = packageVersionSource,
-                      allPackagesPageEntryRevision = listToMaybe packageVersionRevisions
-                    }
-              )
-                . head
-                . sortOn (Down . pkgVersion . pkgId)
+        packageVersions
+          & groupBy ((==) `on` (pkgName . pkgId))
+          & map
+            ( \group ->
+                group
+                  & sortOn (Down . pkgVersion . pkgId)
```

Comment:
TBH I'd appreciate some comments on this pipeline, doing the type inference in my head is hard :D

Reply:
very fair :+1: